statsrelay/src/backend_client/prom_client.rs

```diff
@@ -0,0 +1,522 @@
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+
+use async_trait::async_trait;
+use log::{error, info, warn};
+use prost::Message;
+use reqwest::Client;
+use statsrelay_protobuf::prometheus::metric_metadata::MetricType;
+use statsrelay_protobuf::prometheus::{Label, MetricMetadata, Sample, TimeSeries, WriteRequest};
+use stream_cancel::{Trigger, Tripwire};
+use tokio::select;
+use tokio::sync::{mpsc, Semaphore};
+use tokio::task::JoinHandle;
+use tokio::time::sleep;
+
+use crate::config::PromBackendClientConfig;
+use crate::stats::{self, Counter, Gauge, Scope};
+use crate::statsd_proto::{Event, Id, Owned, Parsed, Pdu};
+
+use super::BackendClient;
+
+#[derive(Clone, Debug)]
+struct Stats {
+    bytes_sent_total: Counter,
+    requests_compression_fail: Counter,
+    requests_fail: Counter,
+    requests_in_flight: Gauge,
+    requests_retry: Counter,
+    requests_total: Counter,
+    samples_parse_fail: Counter,
+    samples_send_fail: Counter,
+    samples_total: Counter,
+}
+
+impl Stats {
+    fn new(scope: Scope) -> Self {
+        Self {
+            bytes_sent_total: scope.counter("bytes_sent_total").unwrap(),
+            requests_compression_fail: scope.counter("requests_compression_fail").unwrap(),
+            requests_fail: scope.counter("requests_fail").unwrap(),
+            requests_in_flight: scope.gauge("requests_in_flight").unwrap(),
+            requests_retry: scope.counter("requests_retry").unwrap(),
+            requests_total: scope.counter("requests_total").unwrap(),
+            samples_parse_fail: scope.counter("samples_parse_fail").unwrap(),
+            samples_send_fail: scope.counter("samples_send_fail").unwrap(),
+            samples_total: scope.counter("samples_total").unwrap(),
+        }
+    }
+}
+
+pub struct PromClient {
+    endpoint: String,
+    inner: Arc<Mutex<PromClientInner>>,
+    sender: mpsc::Sender<Pdu>,
+}
+
+impl PromClient {
+    pub fn new(scope: stats::Scope, endpoint: &str, config: PromBackendClientConfig) -> Self {
+        let inner = Client::builder()
+            .timeout(Duration::from_millis(config.request_timeout_ms))
+            .build()
+            .expect("failed to build client");
```

Comment:
Is there a case in which a caller could recover from a failure to construct this (e.g. by falling back to another backend or something)? In other words, should this return a Result<Self> rather than panicking on construction?

If not, please add a doc-comment to this method to note that it panics on a failure to construct the backing client from a config.